src/asr/asr_chainer.py

```diff
@@ -215,31 +389,64 @@ def train(args):
     with open(args.valid_label, 'rb') as f:
         valid_json = json.load(f)['utts']
 
-    # make minibatch list (variable length)
-    train = make_batchset(train_json, args.batch_size,
-                          args.maxlen_in, args.maxlen_out, args.minibatches)
-    valid = make_batchset(valid_json, args.batch_size,
-                          args.maxlen_in, args.maxlen_out, args.minibatches)
-    # hack to make batchsze argument as 1
-    # actual bathsize is included in a list
-    train_iter = chainer.iterators.SerialIterator(train, 1)
-    valid_iter = chainer.iterators.SerialIterator(
-        valid, 1, repeat=False, shuffle=False)
-
     # prepare Kaldi reader
     train_reader = lazy_io.read_dict_scp(args.train_feat)
     valid_reader = lazy_io.read_dict_scp(args.valid_feat)
 
+    # set up training iterator and updater
+    if ngpu <= 1:
+        # make minibatch list (variable length)
+        train = make_batchset(train_json, args.batch_size,
+                              args.maxlen_in, args.maxlen_out, args.minibatches)
+        # hack to make batchsize argument as 1
+        # actual batchsize is included in a list
+        train_iter = chainer.iterators.SerialIterator(train, 1)
+
+        # set up updater
+        updater = ChainerSeqUpdaterKaldi(
+            train_iter, optimizer, train_reader, gpu_id)
+    else:
+        # set up minibatches
+        train_subsets = []
+        for gid in six.moves.xrange(ngpu):
+            # make subset
+            train_json_subset = {k: v for i, (k, v) in enumerate(train_json.viewitems())
+                                 if i % ngpu == gid}
+            # make minibatch list (variable length)
+            train_subsets += [make_batchset(train_json_subset, args.batch_size,
+                                            args.maxlen_in, args.maxlen_out, args.minibatches)]
+
+        # each subset must have same length for MultiprocessParallelUpdater
+        maxlen = max([len(train_subset) for train_subset in train_subsets])
+        for train_subset in train_subsets:
+            if maxlen != len(train_subset):
+                for i in six.moves.xrange(maxlen - len(train_subset)):
+                    train_subset += [train_subset[i]]
```

Comment:
@sw005320 my last concern is this part. 
If It is no problem, it can be merged.